torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,696 @@ def load_arg(a):
+) -> Optional[torch.dtype]:
+    """
+    Returns the expected dtype of the input and output of this node after
+    convert. If the value is not None, it represents the dtype of the
+    Tensor. If the value is None, it means the value is not a Tensor.
+
+    TODO(future PR, if needed): explicitly spell out the non-Tensor
+    dtypes.
+    """
+    if node.op == 'placeholder':
+        if inputs_seen_counter in input_quantized_idxs:
+            return torch.quint8
+        else:
+            # if dtype is fp32 (default), do nothing
+            # note: other dtypes are not supported
+            return torch.float
+
+    elif node.op in ('call_module', 'call_method', 'call_function'):
+        args_have_no_tensors = \
+            all_node_args_have_no_tensors(
+                node, modules, cache_for_no_tensor_check)
+        if args_have_no_tensors:
+            return None
+
+        # get qconfig to determine the eventual dtype of this node
+        if qconfig is not None:
+            if qhandler is not None and qhandler.input_output_observed():
+                act_dtype, weight_dtype, act_compute_dtype = \
+                    get_qconfig_dtypes(qconfig)
+                return act_dtype
+            else:
+                return torch.float
+        else:
+            return torch.float
+
+    elif node.op == 'get_attr':
+        return torch.float
+
+    elif node.op == 'output':
+        if outputs_seen_counter in output_quantized_idxs:
+            return torch.quint8
+        else:
+            # if dtype is fp32 (default), do nothing
+            # note: other dtypes are not supported
+            return torch.float
+
+    else:
+        raise AssertionError(f'need to handle {node.format_node()}')
+
+def maybe_insert_input_observer_for_arg_or_kwarg_rewrite(
+    node: Union[Node, Any],
+    arg: Argument,
+    qconfig: QConfigAny,
+    load_arg: Callable,
+    model: torch.nn.Module,
+    modules: Dict[str, torch.nn.Module],
+    graph: Graph,
+    node_name_to_target_dtype: Dict[str, Any],
+    qhandler: Optional[QuantizeHandler],
+    prepare_custom_config_dict: Dict[str, Any],
+) -> Argument:
+    """
+    Given a `node` and an `arg`, inserts an input observer between
+    `node` and `arg` if necessary.
+    """
+    # TODO(before land): docblock
+    # print('checking', node, arg)
+
+    # for ops such as torch.cat([x0, x1]),
+    # traverse through the list
+    if isinstance(arg, (list, tuple)):
+        new_arg_to_return = []
+        for inner_arg in arg:
+            new_inner_arg = maybe_insert_input_observer_for_arg_or_kwarg_rewrite(
+                node, inner_arg, qconfig, load_arg, model, modules,
+                graph, node_name_to_target_dtype,
+                qhandler, prepare_custom_config_dict)
+            new_arg_to_return.append(new_inner_arg)
+        return new_arg_to_return
+
+    if not isinstance(arg, Node):
+        return arg
+    assert isinstance(arg, Node)
```

Comment:
nit: is this a bit redundant?

Reply:
yeah, this is just for making mypy happy.  I'll double check if I can remove it.